File: frontend/src/lib/halo.ts
export type HaloCommunitySignal = {
  rating_enabled: boolean;
  rating_average: number | null;
  rating_count: number;
  like_count: number;
  self_reported_visit_count: number;
  proximity_verified_visit_count: number;
  visitability_score: number | null;
  visitability_confidence: number;
};

export type Halo = {
  id: string;
  name: string;
  description: string;
  location_label: string;
  latitude: number;
  longitude: number;
  created_at: string;
  submitted_by: string | null;
  source: string;
  community: HaloCommunitySignal;
  caution: string;
};

export type HaloListResponse = {
  items: Halo[];
  count: number;
};

export type HaloCreateInput = {
  name: string;
  description: string;
  location_label: string;
  latitude: number;
  longitude: number;
  submitted_by?: string;
  rating_enabled?: boolean;
};

export type HaloRatingInput = {
  rating: number;
  liked: boolean;
  visited: boolean;
  visit_latitude?: number;
  visit_longitude?: number;
};

export type HaloRatingResponse = {
  halo: Halo;
  client_id: string;
  rating: number;
  liked: boolean;
  visited: boolean;
  proximity_verified: boolean;
  updated_at: string;
};

type FetchLike = (
  input: string,
  init?: {
    method?: string;
    headers?: Record<string, string>;
    body?: string;
  },
) => Promise<{
  ok: boolean;
  status: number;
  json: () => Promise<unknown>;
}>;

function endpoint(apiBaseUrl: string, path: string): string {
  return `${apiBaseUrl.replace(/\/$/, '')}${path}`;
}

async function expectJson<T>(
  response: Awaited<ReturnType<FetchLike>>,
): Promise<T> {
  if (!response.ok) {
    throw new Error(`Halo request failed with status ${response.status}`);
  }
  return (await response.json()) as T;
}

export async function fetchHalo(
  apiBaseUrl: string,
  options: { query?: string; bbox?: string } = {},
  fetchImpl: FetchLike = fetch,
): Promise<HaloListResponse> {
  const params = new URLSearchParams();
  if (options.query) params.set('q', options.query);
  if (options.bbox) params.set('bbox', options.bbox);
  const query = params.toString();
  const path = `/api/v1/halo${query ? `?${query}` : ''}`;

  return expectJson<HaloListResponse>(
    await fetchImpl(endpoint(apiBaseUrl, path)),
  );
}

export async function createHalo(
  apiBaseUrl: string,
  clientId: string,
  input: HaloCreateInput,
  fetchImpl: FetchLike = fetch,
): Promise<Halo> {
  return expectJson<Halo>(
    await fetchImpl(endpoint(apiBaseUrl, '/api/v1/halo'), {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'X-Client-ID': clientId,
      },
      body: JSON.stringify(input),
    }),
  );
}

export async function rateHalo(
  apiBaseUrl: string,
  clientId: string,
  haloId: string,
  input: HaloRatingInput,
  fetchImpl: FetchLike = fetch,
): Promise<HaloRatingResponse> {
  return expectJson<HaloRatingResponse>(
    await fetchImpl(
      endpoint(apiBaseUrl, `/api/v1/halo/${haloId}/rating`),
      {
        method: 'PUT',
        headers: {
          'Content-Type': 'application/json',
          'X-Client-ID': clientId,
        },
        body: JSON.stringify(input),
      },
    ),
  );
}

export async function setHaloRatingEnabled(
  apiBaseUrl: string,
  clientId: string,
  haloId: string,
  enabled: boolean,
  fetchImpl: FetchLike = fetch,
): Promise<Halo> {
  return expectJson<Halo>(
    await fetchImpl(
      endpoint(apiBaseUrl, `/api/v1/halo/${haloId}/rating-settings`),
      {
        method: 'PUT',
        headers: {
          'Content-Type': 'application/json',
          'X-Client-ID': clientId,
        },
        body: JSON.stringify({ enabled }),
      },
    ),
  );
}

File: backend/code/src/models/halo.py
from datetime import datetime

from pydantic import BaseModel, Field, model_validator


class HaloCreateRequest(BaseModel):
    name: str = Field(min_length=2, max_length=120)
    description: str = Field(min_length=5, max_length=500)
    location_label: str = Field(min_length=2, max_length=160)
    latitude: float = Field(ge=-90, le=90)
    longitude: float = Field(ge=-180, le=180)
    submitted_by: str | None = Field(default=None, max_length=80)
    rating_enabled: bool = True


class HaloRatingRequest(BaseModel):
    rating: int = Field(ge=1, le=5)
    liked: bool = False
    visited: bool = False
    visit_latitude: float | None = Field(default=None, ge=-90, le=90)
    visit_longitude: float | None = Field(default=None, ge=-180, le=180)

    @model_validator(mode="after")
    def coordinates_are_paired(self) -> "HaloRatingRequest":
        if (self.visit_latitude is None) != (self.visit_longitude is None):
            raise ValueError(
                "visit_latitude and visit_longitude must be supplied together"
            )
        return self


class HaloCommunitySignal(BaseModel):
    rating_enabled: bool
    rating_average: float | None = Field(default=None, ge=1, le=5)
    rating_count: int = Field(ge=0)
    like_count: int = Field(ge=0)
    self_reported_visit_count: int = Field(ge=0)
    proximity_verified_visit_count: int = Field(ge=0)
    visitability_score: float | None = Field(default=None, ge=0, le=100)
    visitability_confidence: float = Field(ge=0, le=1)


class Halo(BaseModel):
    id: str
    name: str
    description: str
    location_label: str
    latitude: float
    longitude: float
    created_at: datetime
    submitted_by: str | None = None
    source: str
    community: HaloCommunitySignal
    caution: str = (
        "Halo visitability is community feedback and does not change the "
        "area danger or safety score."
    )


class HaloListResponse(BaseModel):
    items: list[Halo]
    count: int


class HaloRatingResponse(BaseModel):
    halo: Halo
    client_id: str
    rating: int
    liked: bool
    visited: bool
    proximity_verified: bool
    updated_at: datetime



class HaloRatingSettingsRequest(BaseModel):
    enabled: bool
